packages/@aws-containers/aws-ecs-builder/lib/extensions/appmesh.ts

```diff
@@ -0,0 +1,216 @@
+import * as appmesh from '@aws-cdk/aws-appmesh';
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as ecr from '@aws-cdk/aws-ecr';
+import * as ecs from '@aws-cdk/aws-ecs';
+import * as cdk from '@aws-cdk/core';
+import { Service } from '../service';
+import { Container } from './container';
+import { ServiceExtension, ServiceBuild } from './extension-interfaces';
+
+/**
+ * The settings for the App Mesh extension
+ */
+export interface MeshProps {
+  /**
+   * The service mesh into which to register the service
+   */
+  readonly mesh: appmesh.Mesh;
+}
+
+/**
+ * This extension adds an Envoy sidecar to the task definition and
+ * creates the App Mesh resources required to route network traffic
+ * to the container in a service mesh
+ */
+export class AppMeshExtension extends ServiceExtension {
+  protected virtualNode!: appmesh.VirtualNode;
+  protected virtualService!: appmesh.VirtualService;
+  private mesh: appmesh.Mesh;
+
+  constructor(props: MeshProps) {
+    super('appmesh');
+    this.mesh = props.mesh;
+  }
+
+  public prehook(service: Service, scope: cdk.Construct) {
+    this.parentService = service;
+    this.scope = scope;
+
+    // Make sure that the parent cluster for this service has
+    // a namespace attached.
+    if (!this.parentService.cluster.defaultCloudMapNamespace) {
+      this.parentService.cluster.addDefaultCloudMapNamespace({
+        name: 'internal',
+      });
+    }
+  }
+
+  public mutateTaskDefinitionProps(props: ecs.TaskDefinitionProps) {
+    // Find the app extension, to get its port
+    const containerextension = this.parentService.serviceDescription.get('service-container') as Container;
+
+    if (!containerextension) {
+      throw new Error('Appmesh extension requires an application extension');
+    }
+
+    return {
+      ...props,
+
+      // App Mesh requires AWS VPC networking mode so that each
+      // task can have its own IP address
+      networkMode: ecs.NetworkMode.AWS_VPC,
+
+      // This configures the envoy container as a proxy for all
+      // traffic going into and out of the task, with a few exceptions
+      // for metadata endpoints or other ports that need direct
+      // communication
+      proxyConfiguration: new ecs.AppMeshProxyConfiguration({
+        containerName: 'envoy',
+        properties: {
+          appPorts: [containerextension.trafficPort],
+          proxyEgressPort: 15001,
+          proxyIngressPort: 15000,
+          ignoredUID: 1337,
+          ignoredGID: 1338,
+          egressIgnoredIPs: [
+            '169.254.170.2', // Allow services to talk directly to ECS metadata endpoints
+            '169.254.169.254', // and EC2 instance endpoint
+          ],
+          // Note that at some point we will need other extensions like
+          // MySQL to be able to add their ports to this egress ignored
```

Comment:
Ignoring the MySQL port is a workaround for a specific bug in App Mesh (that's about to be resolved), so we can remove that mention.